### PR TITLE
Cache the docs tool to speed up content verification

### PR DESCRIPTION
**Pull request by danielmarbach**
As long as there is no new docstool version pushed this improves the docs tool installation process by 24-25 times.

## Before

image

## After

image

## On cache miss

image

### CODE REVIEW

**Comment by danielmarbach** [comment]
If this works good for us we could also attempt to move it into a composite action.

File: .github/workflows/validate-pull-requests.yml [hunk 1]
```diff
@@ -14,8 +14,19 @@ jobs:
...
+      - name: Check docstool
+        run: |
+           curl https://www.myget.org/F/particular/api/v3/flatcontainer/particular.docstool/index.json >> docstool.version
+           echo "/docstool" >> $env:GITHUB_PATH
+      - name: Cache docstool
+        id: cache-docstool
+        uses: actions/cache@v3
+        with:
+          path: /docstool
+          key: ${{ runner.os }}-docstool-${{ hashFiles('docstool.version') }}
```
Comment (danielmarbach) [inline .github/workflows/validate-pull-requests.yml:26]: I don't think the runner os is really necessary. Should I just use docstool-hash?

File: .github/workflows/validate-pull-requests.yml [hunk 2]
```diff
@@ -14,8 +14,19 @@ jobs:
         uses: actions/setup-dotnet@v3.0.3
         with:
           dotnet-version: 7.0.x
+      - name: Check docstool
+        run: |
+           curl https://www.myget.org/F/particular/api/v3/flatcontainer/particular.docstool/index.json >> docstool.version
```
Comment (danielmarbach) [inline .github/workflows/validate-pull-requests.yml:19]: This file contains all versions. Should I call it .versions?

The get request might slightly more expensive over time when we push more versions but this version is still many times faster than doing nuget list (that one took on average around 10-11 seconds)

File: .github/workflows/validate-pull-requests.yml (hunk 1 above)
Comment (DavidBoike) [reply .github/workflows/validate-pull-requests.yml:26]: Yeah as long as we only run the workflow on one platform it shouldn't matter.

Though the cache keys are a repo-wide thing that would be shared by multiple workflows.

File: .github/workflows/validate-pull-requests.yml [hunk 3]
```diff
@@ -14,8 +14,19 @@ jobs:
...
+           echo "/docstool" >> $env:GITHUB_PATH
```
Comment (DavidBoike) [inline .github/workflows/validate-pull-requests.yml:20]: Is this to add the tool directory to the `PATH`? Is that worth a comment?

File: .github/workflows/validate-pull-requests.yml (hunk 2 above)
Comment (DavidBoike) [reply .github/workflows/validate-pull-requests.yml:19]: How about this?
```suggestion
           curl https://www.myget.org/F/particular/api/v3/flatcontainer/particular.docstool/index.json >> docstool-versions.json
```

And at some point we should clean out some of those versions to conserve MyGet space given I can't think of a single reason to keep very old versions of DocsTool around.

File: .github/workflows/validate-pull-requests.yml [hunk 4]
```diff
@@ -14,8 +14,19 @@ jobs:
...
+      - name: Cache docstool
+        id: cache-docstool
+        uses: actions/cache@v3
```
Comment (DavidBoike) [inline .github/workflows/validate-pull-requests.yml:23]:
```suggestion
        uses: actions/cache@v3.3.1
```

File: .github/workflows/verify-master.yml [hunk 5]
```diff
@@ -17,8 +17,19 @@ jobs:
         uses: actions/setup-dotnet@v3.0.3
         with:
           dotnet-version: 7.0.x
+      - name: Check docstool
+        run: |
+           curl https://www.myget.org/F/particular/api/v3/flatcontainer/particular.docstool/index.json >> docstool.version
+           echo "/docstool" >> $env:GITHUB_PATH
+      - name: Cache docstool
+        id: cache-docstool
+        uses: actions/cache@v3
```
Comment (DavidBoike) [inline .github/workflows/verify-master.yml:26]:
```suggestion
        uses: actions/cache@v3.3.1
```

**Review by DavidBoike — APPROVED** [review]

File: .github/workflows/validate-pull-requests.yml (hunk 3 above)
Comment (danielmarbach) [reply .github/workflows/validate-pull-requests.yml:20]: I first tried to use $HOME/.dotnet/tools but somehow couldn't get it to work and then switched to using a dedicated path. I guess if we could use the home tools path there would be nothing need to add to the path but we might be caching more than we want. If we leave it as is adding a comment might be good yes.

File: .github/workflows/validate-pull-requests.yml (hunk 1 above)
Comment (danielmarbach) [reply .github/workflows/validate-pull-requests.yml:26]: Yeah they are but that should be a problem. I assume populating the cache is idempotent. So we even get bigger caching benefits

Comment (danielmarbach) [inline .github/workflows/validate-pull-requests.yml:26]:
```suggestion
          key: docstool-${{ hashFiles('docstool-versions.json') }}
```

**Comment by danielmarbach** [comment]
Concurrent cache update works like a charm

image

**Pull request merged**